### PR TITLE
Kj/demo

### PR DESCRIPTION
**Pull request by Katinka666**

### CODE REVIEW

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 1]
```diff
@@ -0,0 +1,846 @@
...
+The _temperature-provider2_ system that we have just registered provides three services in the Local Cloud:
+- **Kelvin info:** provides temperature information using the Kelvin scale,
+- **Celsius info:** provides temperature information using the Celsius scale,
+- **alert service:** sends an alert if the temperature is extreme (by default, these thresholds are 10 and 25 Celsius, but the consumer can overwrite them).
+
+We have to registrer these services one by one.
+
+**1. Kelvin info:** 
+We have to provide the following information:
+- **service definition name:** We have to use the name of one of the existing service definitions stored in the Local Cloud. In this example, this is kelvin-info.
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:212]: This is not true. If the service definition not exist, the registration process will create them.

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 2]
```diff
@@ -0,0 +1,846 @@
...
+- **version:** We will use the default version, so we can leave this field blank.
+- **expires at:** This is a timestamp in the future, when the service is no longer funtioning. For Kelvin info, we set this to 01. 01. 2030. 
+- **metadata:** This can be customised depending on the service. For temperature information, we define the margin of error, which is 0,5 degree.
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:215]: you should use dot instead of comma.

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 3]
```diff
@@ -0,0 +1,846 @@
...
+- **interfaces:** All the services use http protocol, so we will go with the template named generic-http, that already exists in the Local Cloud. Note that in our case, the service discovery interface policy is set to _restricted_, which means that only already existing interface templates can be used. If you set this to _extendable_ or _open_, you can use non-existent interface templates, and they will be created as well. The interface provided by the Kelvin info service is the following:
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:216]: http protocol => HTTP protocol

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 4]
```diff
@@ -0,0 +1,846 @@
...
+  "interfaces": [
+    {
+      "templateName": "generic-http",
+      "protocol": "http",
+      "policy": "NONE",
+      "properties": {
+        "accessAddresses": ["192.168.56.116", "tp2.greenhouse.com"],
+        "accessPort": 8080,
+        "basePath": "/kelvin",
+        "operations": {"query-temperature": { "method": "get", "path": "/query"} }
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:243]: should be GET instead of get

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 5]
```diff
@@ -0,0 +1,846 @@
...
+  "interfaces": [
+    {
+      "templateName": "generic-http",
+      "protocol": "http",
+      "policy": "NONE",
+      "properties": {
+        "accessAddresses": ["192.168.56.116", "tp2.greenhouse.com"],
+        "accessPort": 8080,
+        "basePath": "/celsius",
+        "operations": {"query-temperature": { "method": "get", "path": "/query"} }
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:277]: should be GET instead of get

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 6]
```diff
@@ -0,0 +1,846 @@
...
+      }
+    }
+  ]
+}'
+~~~
+
+**2. Alert service:**
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:284]: this should be 3 instead of 2

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 7]
```diff
@@ -0,0 +1,846 @@
...
+~~~
+
+**2. Alert service:** 
+Our last service will be responsible for sending error messages. The registration data is the following:
+- **service definition name:** In this case this is alert-service.
+- **version:** We will use the default version.
+- **expires at:** The alert service expires a bit earlier than the previous ones, so we set this to 01. 01. 2025.
+- **metadata:** For alert service, the maximum possible delay is given, which is 15 sec.
+- **interfaces:** The interfaces provided by the alert service are the following:
+  - GET tp2.greenhouse.com:8000/alert/subscribe
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:291]: A subscribe should be POST

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 8]
```diff
@@ -0,0 +1,846 @@
...
+
+**2. Alert service:** 
+Our last service will be responsible for sending error messages. The registration data is the following:
+- **service definition name:** In this case this is alert-service.
+- **version:** We will use the default version.
+- **expires at:** The alert service expires a bit earlier than the previous ones, so we set this to 01. 01. 2025.
+- **metadata:** For alert service, the maximum possible delay is given, which is 15 sec.
+- **interfaces:** The interfaces provided by the alert service are the following:
+  - GET tp2.greenhouse.com:8000/alert/subscribe 
+  - GET tp2.greenhouse.com:8000/alert/unsubscribe
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:292]: An unsubscribe should be DELETE

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 9]
```diff
@@ -0,0 +1,846 @@
...
+}'
+~~~
+
+**2. Alert service:** 
+Our last service will be responsible for sending error messages. The registration data is the following:
+- **service definition name:** In this case this is alert-service.
+- **version:** We will use the default version.
+- **expires at:** The alert service expires a bit earlier than the previous ones, so we set this to 01. 01. 2025.
+- **metadata:** For alert service, the maximum possible delay is given, which is 15 sec.
+- **interfaces:** The interfaces provided by the alert service are the following:
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:290]: There is no multiple interfaces (as we understand interface in Arrowhead). So I'd rephrase this sentence: The interface's endpoints provided by ...

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 10]
```diff
@@ -0,0 +1,846 @@
...
+      "protocol": "http",
+      "policy": "NONE",
+      "properties": {
+        "accessAddresses": ["192.168.56.116", "tp2.greenhouse.com"],
+        "accessPort": 8080,
+        "basePath": "/alert",
+        "operations": {
+          "subscribe": { "method": "get", "path": "/subscribe"},
+          "unsubscribe": { "method": "get", "path": "/unsubscribe"},
+          "set-threshold": { "method": "post", "path": "/threshold"}
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:322]: change the methods, also make them uppercase

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 11]
```diff
@@ -0,0 +1,846 @@
...
+      "properties": {
+        "accessAddresses": ["192.168.56.116", "tp2.greenhouse.com"],
+        "accessPort": 8080,
+        "basePath": "/kelvin",
+        "operations": {"query-temperature": { "method": "get", "path": "/query"} }
+      }
+    }
+  ]
+}'
+~~~
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:248]: missing response

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 12]
```diff
@@ -0,0 +1,846 @@
...
+      "properties": {
+        "accessAddresses": ["192.168.56.116", "tp2.greenhouse.com"],
+        "accessPort": 8080,
+        "basePath": "/celsius",
+        "operations": {"query-temperature": { "method": "get", "path": "/query"} }
+      }
+    }
+  ]
+}'
+~~~
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:282]: missing response

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 13]
```diff
@@ -0,0 +1,846 @@
...
+        "operations": {
+          "subscribe": { "method": "get", "path": "/subscribe"},
+          "unsubscribe": { "method": "get", "path": "/unsubscribe"},
+          "set-threshold": { "method": "post", "path": "/threshold"}
+        }
+      }
+    }
+  ]
+}'
+~~~
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:328]: missing response

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 14]
```diff
@@ -0,0 +1,846 @@
...
+                "path": "/subscribe",
+                "method": "GET"
+              },
+              "unsubscribe": {
+                "path": "/unsubscribe",
+                "method": "GET"
+              },
+              "set-threshold": {
+                "path": "/threshold",
+                "method": "POST"
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:431]: Don't forget to change the methods (after you changed them in the registration part)

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 15]
```diff
@@ -0,0 +1,846 @@
...
+  ],
+  "count": 1
+}
+~~~
+
+Since there was exactly one response, we learned that the service instance id is _temperature-provider2::alert-service::1.0.0_. We can revoke the service by sending the following request:
+
+~~~
+curl -X 'DELETE' \
+  'http://localhost:8443/serviceregistry/service-discovery/revoke/temperature-provider2%3A%3Aalert-service%3A%3A1.0.0' \
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:450]: maybe some note why you have to use the %3A-s in the url.

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 16]
```diff
@@ -0,0 +1,846 @@
...
+  -H 'Authorization: Bearer SYSTEM//temperature-provider2'
+~~~
+
+This operation only works, if the service we are deleting is associated with our system. Otherwise, the following error message will be received:
+~~~
+{
+  "errorMessage": "Revoking other systems' service is forbidden",
+  "errorCode": 403,
+  "exceptionType": "FORBIDDEN",
+  "origin": "DELETE /serviceregistry/service-discovery/revoke/{instanceId}"
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:461]: Looks like you catch a minor bug. The origin (in case of using path parameters) should contain the actual value of the instanceId, not the literal {instanceId}.

Please fix the demo with the actual value, and create an issue for the bug.

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 17]
```diff
@@ -0,0 +1,846 @@
...
+The Service Registry will delete our system.
+
+## Example 2: Consumer
+
+In this example, we'll see how:
+- a **system** called _temperature-consumer-1_, 
+- running on the **device** named _weather-displayer1_,
+- looks up for available _Kelvin-info_ **services**,
+
+after registering itself and it's device into the Local Cloud.
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:484]: typo it's => its

File: docs/serviceregistry/manuals/temperature-http-demo.md [hunk 18]
```diff
@@ -0,0 +1,846 @@
...
+  -H 'Authorization: Bearer SYSTEM//temperature-consumer1'
+~~~
+
+Note that this operation will only be successful, if no system is connected to the device anymore. Otherwise you will get the following error message:
+~~~
+{
+  "errorMessage": "At least one system is assigned to this device.",
+  "errorCode": 423,
+  "exceptionType": "LOCKED",
+  "origin": "DELETE /serviceregistry/device-discovery/revoke/{name}"
```
Comment (rbocsi) [inline docs/serviceregistry/manuals/temperature-http-demo.md:554]: fix the origin here (and later we will fix in the code)

**Review by rbocsi — CHANGES_REQUESTED** [review]
I really like this demo. However, I found some small issues.

**Review by rbocsi — APPROVED** [review]

**Review by borditamas — APPROVED** [review]

**Comment by borditamas** [comment]
átemeltem az mkdocs site-ba, ezt a PR-t nem kell mergelni
